### PR TITLE
CCIT-90: Virto Commerce 3.x Module template

### PR DESCRIPTION
**Pull request by mvktsk**
Created Virto Commerce 3.x Module template

Template marked as preview.
To give possibility install V2 and V3 at the same time ProductID changed Module.Installer project and  TemplateID added In  Module.Template project.

Known issues; in solution projects placed in src and test folders but on disk used old folder structure.

### CODE REVIEW

**Review by eMazeika — CHANGES_REQUESTED** [review]
1. fails to build if project name contains "."
2. 1 module template only? :)

**Comment by mvktsk** [comment]
> 1. fails to build if project name contains "."
As designed 
> 2. 1 module template only? :)
As designed 
:)

File: CSharp/VirtoCommerce.Module.Template/src/Core/Core.csproj [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="VirtoCommerce.Platform.Core" Version="3.0.0-rc4-0180" />
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.Module.Template/src/Core/Core.csproj]: reference latest package version

File: CSharp/VirtoCommerce.Module.Template/src/Data/Data.csproj [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    <Folder Include="Repositories\" />
+    <Folder Include="Services\" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="3.1.2">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+    <PackageReference Include="VirtoCommerce.Platform.Data" Version="3.0.0-rc4-0180" />
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.Module.Template/src/Data/Data.csproj]: reference latest package version

File: CSharp/VirtoCommerce.Module.Template/src/Web/Module.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
+using $ext_safeprojectname$.Core;
+using $ext_safeprojectname$.Data.Repositories;
+using VirtoCommerce.Platform.Core.Modularity;
+using VirtoCommerce.Platform.Core.Security;
+using VirtoCommerce.Platform.Core.Settings;
+using System.Linq;
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.Module.Template/src/Web/Module.cs:1]: move to be line 1

File: CSharp/VirtoCommerce.Module.Template/src/Web/Module.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
+using $ext_safeprojectname$.Core;
+using $ext_safeprojectname$.Data.Repositories;
+using VirtoCommerce.Platform.Core.Modularity;
+using VirtoCommerce.Platform.Core.Security;
+using VirtoCommerce.Platform.Core.Settings;
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.Module.Template/src/Web/Module.cs]: sort alphabetically

File: CSharp/VirtoCommerce.Module.Template/src/Web/Scripts/module.js [hunk 5]
```diff
@@ -26,16 +26,16 @@ angular.module(moduleName, [])
...
+        function (mainMenuService, widgetService, $state) {
             //Register module in main menu
             var menuItem = {
                 path: 'browse/$ext_safeprojectnamecamel$',
                 icon: 'fa fa-cube',
                 title: '$ext_projectname$',
                 priority: 100,
                 action: function () { $state.go('workspace.$ext_supersafenamejs$State'); },
-                permission: '$ext_safeprojectnamecamel$.WebPermission'
+                permission: '$ext_safeprojectnamecamel$:read'
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.Module.Template/src/Web/Scripts/module.js]: incorrect permission generated

File: CSharp/VirtoCommerce.Module.Template/src/Web/module.manifest [hunk 6]
```diff
@@ -2,32 +2,26 @@
...
+        <author>$username$</author>
     </authors>
     <owners>
       <owner>$registeredorganization$</owner>
     </owners>
-  
+
     <releaseNotes>First version.</releaseNotes>
     <copyright>Copyright © $year$ $registeredorganization$. All rights reserved</copyright>
     <tags>extension module</tags>
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.Module.Template/src/Web/module.manifest:22]: these are missing in powershell template. Need to synch with it: add or remove.

**Review by eMazeika — CHANGES_REQUESTED** [review]

File: CSharp/VirtoCommerce.Module.Template/src/Core/Core.csproj (hunk 1 above)
Comment (mvktsk) [reply CSharp/VirtoCommerce.Module.Template/src/Core/Core.csproj]: fixed

File: CSharp/VirtoCommerce.Module.Template/src/Data/Data.csproj (hunk 2 above)
Comment (mvktsk) [reply CSharp/VirtoCommerce.Module.Template/src/Data/Data.csproj]: fixed

File: CSharp/VirtoCommerce.Module.Template/src/Web/module.manifest (hunk 6 above)
Comment (mvktsk) [reply CSharp/VirtoCommerce.Module.Template/src/Web/module.manifest:22]: Fixed

File: CSharp/VirtoCommerce.Module.Template/src/Web/Module.cs (hunk 4 above)
Comment (mvktsk) [reply CSharp/VirtoCommerce.Module.Template/src/Web/Module.cs]: Fixed

File: CSharp/VirtoCommerce.Module.Template/src/Web/Module.cs (hunk 3 above)
Comment (mvktsk) [reply CSharp/VirtoCommerce.Module.Template/src/Web/Module.cs:1]: Fixed

File: CSharp/VirtoCommerce.Module.Template/src/Web/Scripts/module.js (hunk 5 above)
Comment (mvktsk) [reply CSharp/VirtoCommerce.Module.Template/src/Web/Scripts/module.js]: fixed

File: CSharp/VirtoCommerce.Module.Template/src/Web/Scripts/module.js [hunk 7]
```diff
@@ -26,16 +26,16 @@ angular.module(moduleName, [])
...
+        function (mainMenuService, widgetService, $state) {
             //Register module in main menu
             var menuItem = {
                 path: 'browse/$ext_safeprojectnamecamel$',
                 icon: 'fa fa-cube',
                 title: '$ext_projectname$',
                 priority: 100,
                 action: function () { $state.go('workspace.$ext_supersafenamejs$State'); },
-                permission: '$ext_safeprojectnamecamel$.WebPermission'
+                permission: '$ext_supersafenamejs$:read'
```
Comment (eMazeika) [inline CSharp/VirtoCommerce.Module.Template/src/Web/Scripts/module.js]: should be :access, not :read

Comment (mvktsk) [reply CSharp/VirtoCommerce.Module.Template/src/Web/Scripts/module.js]: fixed

**Review by eMazeika — APPROVED** [review]

**Pull request merged**